Add extra check for control plane ingress pods and containers

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'test', 'authentication'

**Context**: 

Authentication tests are flaky.
It looks like when we invoke these tests, pods are ready but the containers are not.

Here is a sample output of pods status vs running container status.

```
tests/post/steps/test_authentication.py:133: Failed
----------------------------- Captured stdout call -----------------------------
NAMESPACE          NAME                                             READY   STATUS    RESTARTS   AGE
metalk8s-ingress   nginx-ingress-control-plane-controller-ptr56     0/1     Running   0          14s
metalk8s-ingress   nginx-ingress-controller-t9gr7                   1/1     Running   0          27s
metalk8s-ingress   nginx-ingress-default-backend-69558976b7-75ftd   1/1     Running   0          51s
```

**Summary**:

Ensure that we check the container status for Nginx-ingress-control plane pods before running authentication tests else we will result in flaky tests.

Increase the delay between a failed tests and a fired retry so that requests lib does not complain of handling an exception while another exception was triggered.



**Acceptance criteria**: 

All authentication tests should pass
Less falky


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

alexandre-allard commented on tests/post/steps/test_authentication.py:
```diff
@@ -81,6 +81,50 @@ def control_plane_ip(host):
 # Given {{{
 
+
+@given(parsers.parse(
```
`parsers.parse` is not needed if there is nothing to parse :)

alexandre-allard commented on tests/post/steps/test_authentication.py:
```diff
@@ -81,6 +81,50 @@ def control_plane_ip(host):
+
+@given(parsers.parse(
+    "the control-plane Ingress container is Ready"))
+def check_cp_ingress_containers(request, host, k8s_client):
```
Can't you use the same `given` you use for `And pods with label 'app.kubernetes.io/name=dex' are 'Ready'` ? To me, it looks like the same.
```
And pods with label 'release=nginx-ingress-control-plane' are 'Ready'
```
edit: nvm, didn't see that you check the container status here, not pod one, I though it was the same (so a pod can be ready and not the containers inside ?).
Still, this function could be variabilized (is that even a word ?) to do a `check_pod_containers_status`, rather than this very specific use case.

Ebaneck replied:
The extra check for container status has been added to the same function that checks ingress pods.
so now the function has the following submethods
- check ingress pod status
- obtain the full name of ingress pods
- then finally check the status of containers in that pod

alexandre-allard commented on tests/post/features/authentication.feature:
```diff
@@ -11,6 +11,7 @@ Feature: Authentication is up and running
 
     Scenario: Reach the OpenID Config
         Given the Kubernetes API is available
+        And the control-plane Ingress container is Ready
```
Shouldn't we add this line in each and every scenarios below ?
Maybe, we can even add a `Background` at the beginning of the feature, since we always need the same things:
```
    Background:
        Given the Kubernetes API is available
        And the control-plane Ingress container is Ready
        And the control-plane Ingress path '/oidc' is available
        And pods with label 'app.kubernetes.io/name=dex' are 'Ready'

    Scenario: Reach the OpenID Config
        Then we can reach the OIDC openID configuration

    Scenario: Access HTTPS service
        When we perform a request on '/oidc/' with port '8443' on control-plane IP
        Then the server returns '404' with message '404 page not found'

    Scenario: Login to Dex using incorrect email
        When we login to Dex as 'admin@metalk8s.com' using password 'password'
        Then authentication fails with login error

    Scenario: Login to Dex using correct email and password
        When we login to Dex as 'admin@metalk8s.invalid' using password 'password'
        Then the server returns '303' with an ID token
```
see: https://pytest-bdd.readthedocs.io/en/latest/#backgrounds

Ebaneck replied:
Agreed that we could improve the entire tests suite and not limited to authentication by adding `Background`  scenarios.

Worth creating an issue to track this refactoring.
I have added the line to each scenario for now

Ebaneck replied:
So instead of adding the `Given` to each line, I merged the check for container status in the same function that checks for pod status using submethods

gdemonet replied:
Let's create this tracking issue then @Ebaneck

alexandre-allard commented on tests/post/steps/test_authentication.py:
```diff
@@ -81,25 +81,64 @@ def control_plane_ip(host):
 
+
 @given(parsers.parse(
     "the control-plane Ingress path '{path}' is available"))
```
What this `path` is used for ? Does not seem to be used anywhere.

Ebaneck replied:
True... here is why: https://github.com/scality/metalk8s/pull/2213/files#diff-6065c212c4eed6c66088271ca7d2eddfR142

gdemonet commented on tests/post/features/authentication.feature:
```diff
@@ -12,26 +12,30 @@ Feature: Authentication is up and running
     Scenario: Reach the OpenID Config
         Given the Kubernetes API is available
         And the control-plane Ingress path '/oidc' is available
```
If I may suggest something, we have a "Ingress-specific `Given`", which is supposed to make sure the specified path is available. This check could perform all the required checks to move forward, including the container check:

- First, check that the `nginx-ingress-control-plane` DaemonSet is OK (there's a Pod running on all `master` Nodes, or maybe at least one, not sure)
- Then, check that for each Pod belonging in this DaemonSet, all of its containers are Ready
- Then, through one of the available control plane IPs (in our case, there's only the Bootstrap's that should work), try to connect to `https://<cp-ip>:8443/<path>`, with `<path>` being parsed in the `Given`

If all of these succeed (the last check should just validate that we do not receive a 502 or a 404 from the default backend), then any following query should work. And there's no need to _know_ in advance which Pod make up the service that the Ingress object is pointing to.

Ebaneck replied:
I think point 1 & 2 have been addressed in this PR maybe not word for word.

I am going to open a second PR to address point 3 separately

gdemonet commented on tests/post/steps/test_authentication.py:
```diff
@@ -254,4 +287,16 @@ def _dex_auth_request(control_plane_ip, username, password):
+        pytest.fail("Unable to get pod name with error: {}".format(exc))
+
+    for pod in pods:
+        return pod.metadata.name
```
Not fond of this. It's misleading, as you're getting "the name of the first Pod that matched this query", not "the Pod name". Why not check all Pods?

gdemonet commented on tox.ini:
```diff
@@ -130,6 +130,7 @@ markers =
     ingress: tag a BDD feature as related to ingress
     volume: tag a BDD feature as related to Volume management
     bootstrap: tag a BDD feature as related to bootstrap
+    authentication: tag a BDD feature as related to authentication
```
Good point, however this should have its dedicated commit (not related to your changes above).

gdemonet reviewed (changes requested):
The added checks are OK but could benefit from some clean-up, at least to make the flow obvious for readers. You could open a follow-up PR, since this one is supposed to reduce flakiness.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

Ebaneck committed:
authentication: add extra check for nginx-ingress-controller container status

Initially, we didn't check the status of containers for pods in
nginx-ingress-control-plane. As such we triggered tests while container
state was unready making most of the tests to fail in the CI.

Now, we ensure to check the status of pods and their corresponding
containers before launching tests.

Ebaneck committed:
tests: add markers for authentication tests

Initially, the CI complaints of PytestUnknownMarkWarning because
the authentication BDD feature is not grouped to run with a specific marker.

Now, we tag the authentication feature making it possible to run the sets of
feature using `tox -e tests -- -m "authentication"`

alexandre-allard commented on tests/post/steps/test_authentication.py:
```diff
@@ -81,26 +81,45 @@ def control_plane_ip(host):
+            pytest.fail("unable to get pods with error: {}".format(exc))
+
+        assert len(pods) > 0, (
+            "Expected only one Pod with label {}, but found {}"
```
For me, the assertion and the message are not consistent, either we have:
```
assert pods, "No Pod with label {} found".format(label)
```
Or:
```
assert len(pods) == 1, (
    "Expected only one Pod with label {}, but found {}"
).format(label, len(pods))
```

gdemonet replied:
Unfortunately, you should have picked the first option. We can't know how many Pods we'll find, as it depends on the number of `master` Nodes in the cluster.

gdemonet reviewed (dismissed):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

Ebaneck commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
